src/NLog/Contexts/GlobalContext.cs

```diff
@@ -0,0 +1,220 @@
+// modification, are permitted provided that the following conditions 
+// are met:
+// 
+// * Redistributions of source code must retain the above copyright notice, 
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Contexts
+{
+    using NLog.Internal;
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Global Context - a dictionary structure to hold per-application-instance values.
+    /// </summary>
+    public class GlobalContext : IContext
+    {
+        private static IContext currentInstance = null;
+        private static readonly object syncRoot = new object();
+        private Dictionary<string, object> dict;
+        private HashSet<String> keys;
+
+        private GlobalContext()
+        {
+            this.dict = new Dictionary<string, object>();
+            keys = new HashSet<string>();
+        }
+
+        /// <summary>
+        /// Current instance of the context.
+        /// </summary>
+        public static IContext Instance
+        {
+            get
+            {
+                if (currentInstance == null)
+                {
+                    lock (syncRoot)
+                    {
+                        if (currentInstance == null)
+                        {
+                            currentInstance = new GlobalContext();
+                        }
+                    }
+                }
+                return currentInstance;
+            }
+        }
+
+        /// <summary>
+        /// Returns the keys in the context.
+        /// </summary>
+        public HashSet<string> Keys
+        {
+            get
+            {
+                return keys;
```

Comment:
proposal: ` return new HashSet<string>(dict.Keys);` and remove the "shadow collection"